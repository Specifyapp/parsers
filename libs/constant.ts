import { TokensType } from '../types';

export const tokenTypesList: Array<TokensType> = [
  'border',
  'color',
  'gradient',
  'duration',
  'measurement',
  'opacity',
  'shadow',
  'textStyle',
  'depth',
];
