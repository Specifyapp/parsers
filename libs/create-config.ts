import { Parser, Token } from '../types';

type UnknownParser<F> = { options: unknown; fn: F; name: string };
type ReturnPipe<T> = {
  run: (tokens: Array<Token>) => T;
  build: () => unknown;
};

export function createConfig<A extends unknown[], B>(
  ab: UnknownParser<(...a: A) => B>,
): ReturnPipe<B>;
export function createConfig<A extends unknown[], B, C>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
): ReturnPipe<C>;
export function createConfig<A extends unknown[], B, C, D>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
): ReturnPipe<D>;
export function createConfig<A extends unknown[], B, C, D, E>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
): ReturnPipe<E>;
export function createConfig<A extends unknown[], B, C, D, E, F>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
  ef: UnknownParser<(e: E) => F>,
): ReturnPipe<F>;
export function createConfig<A extends unknown[], B, C, D, E, F, G>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
  ef: UnknownParser<(e: E) => F>,
  fg: UnknownParser<(f: F) => G>,
): ReturnPipe<G>;
export function createConfig<A extends unknown[], B, C, D, E, F, G, H>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
  ef: UnknownParser<(e: E) => F>,
  fg: UnknownParser<(f: F) => G>,
  gh: UnknownParser<(g: G) => H>,
): ReturnPipe<H>;
export function createConfig<A extends unknown[], B, C, D, E, F, G, H, I>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
  ef: UnknownParser<(e: E) => F>,
  fg: UnknownParser<(f: F) => G>,
  gh: UnknownParser<(g: G) => H>,
  hi: UnknownParser<(h: H) => I>,
): ReturnPipe<I>;
export function createConfig<A extends unknown[], B, C, D, E, F, G, H, I, J>(
  ab: UnknownParser<(...a: A) => B>,
  bc: UnknownParser<(b: B) => C>,
  cd: UnknownParser<(c: C) => D>,
  de: UnknownParser<(d: D) => E>,
  ef: UnknownParser<(e: E) => F>,
  fg: UnknownParser<(f: F) => G>,
  gh: UnknownParser<(g: G) => H>,
  hi: UnknownParser<(h: H) => I>,
  ij: UnknownParser<(i: I) => J>,
): ReturnPipe<J>;

export function createConfig(...parsers: Array<UnknownParser<(...params: unknown[]) => unknown>>): {
  run: (tokens: Array<Token>) => unknown;
  build: () => unknown;
} {
  return {
    build: () =>
      parsers.map(parser => {
        const step: Parser = {
          name: parser.name,
        };
        if (parser.options) {
          step.options = parser.options;
        }
        return step;
      }),
    run: (tokens: Array<Token>): unknown => {
      return parsers.reduce<unknown>(async (value, parser) => parser.fn(await value), tokens);
    },
  };
}
