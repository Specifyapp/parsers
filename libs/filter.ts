import { AtLeast, TokensType, TokenTypeFromTokenTypeName } from '../types';

export function filter<T extends TokensType>(
  tokens: Array<AtLeast<{ type: TokensType }>>,
  type: T,
) {
  return tokens.filter(token => token.type === type) as Array<TokenTypeFromTokenTypeName<T>>;
}
