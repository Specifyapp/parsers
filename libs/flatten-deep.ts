export function flattenObject(obj: Record<string, any>) {
  const flattened: Record<string, any> = {};

  Object.keys(obj).forEach(key => {
    if (typeof obj[key] === 'object' && obj[key] !== null) {
      Object.assign(flattened, flattenObject(obj[key]));
    } else {
      flattened[key] = obj[key];
    }
  });

  return flattened;
}
