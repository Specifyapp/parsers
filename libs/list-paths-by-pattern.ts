import { has, get } from 'lodash';

export default function listPathsByPattern<T extends object>(
  base: T | Array<T>,
  pattern: string,
): Array<string> {
  try {
    if (!pattern.includes('[*]')) return has(base, pattern) ? [pattern] : [];
    return pattern
      .split(/(?=\[\*])|(?<=\[\*])/g) // split without remove the delimiter
      .reduce<Array<string>>(
        (paths, item) => {
          if (item === '[*]') {
            return paths
              .map(basePath => {
                const iteration = basePath ? get(base, basePath) : base;
                if (!Array.isArray(iteration)) return [];
                return Array(iteration.length)
                  .fill(0)
                  .reduce(
                    (acc, _, index) =>
                      has(base, `${basePath}[${index}]`) ? [...acc, `${basePath}[${index}]`] : acc,
                    [],
                  );
              })
              .flat(2);
          } else {
            return paths.reduce<Array<string>>(
              (acc, elm) => (has(base, `${elm}${item}`) ? [...acc, `${elm}${item}`] : acc),
              [],
            );
          }
        },
        [''],
      );
  } catch (err) {
    throw err;
  }
}
