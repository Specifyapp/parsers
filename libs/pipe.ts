type UnknownFunction = (...params: unknown[]) => unknown;

// Copied from https://github.com/gcanti/fp-ts/blob/1.15.0/src/function.ts#L209 with some
// modifications to make the first function allow multiple parameters.
export function pipe<A extends unknown[], B>(ab: (this: void, ...a: A) => B): (...args: A) => B;
export function pipe<A extends unknown[], B, C>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
): (...args: A) => C;
export function pipe<A extends unknown[], B, C, D>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
): (...args: A) => D;
export function pipe<A extends unknown[], B, C, D, E>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
): (...args: A) => E;
export function pipe<A extends unknown[], B, C, D, E, F>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
  ef: (this: void, e: E) => F,
): (...args: A) => F;
export function pipe<A extends unknown[], B, C, D, E, F, G>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
  ef: (this: void, e: E) => F,
  fg: (this: void, f: F) => G,
): (...args: A) => G;
export function pipe<A extends unknown[], B, C, D, E, F, G, H>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
  ef: (this: void, e: E) => F,
  fg: (this: void, f: F) => G,
  gh: (this: void, g: G) => H,
): (...args: A) => H;
export function pipe<A extends unknown[], B, C, D, E, F, G, H, I>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
  ef: (this: void, e: E) => F,
  fg: (this: void, f: F) => G,
  gh: (this: void, g: G) => H,
  hi: (this: void, h: H) => I,
): (...args: A) => I;
export function pipe<A extends unknown[], B, C, D, E, F, G, H, I, J>(
  ab: (this: void, ...a: A) => B,
  bc: (this: void, b: B) => C,
  cd: (this: void, c: C) => D,
  de: (this: void, d: D) => E,
  ef: (this: void, e: E) => F,
  fg: (this: void, f: F) => G,
  gh: (this: void, g: G) => H,
  hi: (this: void, h: H) => I,
  ij: (this: void, i: I) => J,
): (...args: A) => J;

export function pipe(...fns: UnknownFunction[]): UnknownFunction {
  return (...initialParams: unknown[]): unknown =>
    fns.reduce<unknown>((value, fn, index) => {
      const params = index === 0 ? (value as unknown[]) : [value];
      return fn(...params);
    }, initialParams);
}
