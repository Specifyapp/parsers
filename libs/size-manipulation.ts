import { MeasurementValue } from '../types';

function convertMeasurement(
  from: MeasurementValue,
  toUnit: 'px' | 'rem',
  basePixelValue?: number | undefined,
): MeasurementValue;
function convertMeasurement(
  from: string,
  toUnit: 'px' | 'rem',
  basePixelValue?: number | undefined,
): string;
function convertMeasurement(
  from: string | MeasurementValue,
  toUnit: 'px' | 'rem',
  basePixelValue = 16,
): MeasurementValue | string {
  if (!['px', 'rem'].includes(toUnit)) throw new Error('Unknown size unit');
  let fromValue = typeof from === 'object' && 'measure' in from ? from.measure : parseFloat(from);
  const fromUnit = typeof from === 'string' ? from.match(/[\d.\-\+]*\s*(.*)/)![1] || '' : from.unit;
  let pxValue;
  if (fromUnit === 'px') {
    pxValue = fromValue;
  } else if (fromUnit === 'rem') {
    pxValue = fromValue * basePixelValue;
  }
  if (!pxValue) throw new Error('Unknown size unit');

  const result = {
    unit: toUnit === 'px' ? 'px' : 'rem',
    measure: toUnit === 'px' ? pxValue : pxValue / basePixelValue,
  };
  return typeof from === 'string' ? `${result.measure}${result.unit}` : result;
}

export default convertMeasurement;
