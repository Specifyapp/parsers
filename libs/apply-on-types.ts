import { tokenTypesList } from './constant';
import { TokensType } from '../types';

export function getTokenTypesToApplyFn<T extends { filter?: { types?: Array<TokensType> } }>(
  options: T,
) {
  if (!options?.filter?.types) {
    return tokenTypesList;
  }
  return tokenTypesList.filter(tokenType => options.filter!.types!.includes(tokenType));
}
