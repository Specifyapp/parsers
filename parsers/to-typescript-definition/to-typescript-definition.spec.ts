import seeds from '../../tests/seeds';
import toTypescriptDefinition, { InputDataType } from './to-typescript-definition.parser';
import libs from '../global-libs';

const tokens = seeds().tokens;

describe('Tokens to Typescript types definitions', () => {
  it('Should generate all the types with default formatting', async () => {
    const result = await toTypescriptDefinition(tokens as InputDataType, {}, libs);

    expect(result).toMatchSnapshot();
  });

  it('Should generate all the types with correct type name', async () => {
    const result = await toTypescriptDefinition(
      tokens as InputDataType,
      {
        typesNamesMapping: {
          color: 'helloColor',
          vector: 'myVector',
          textStyle: 'text_style',
        },
      },
      libs,
    );

    expect(result.includes("export type helloColor = 'colors-accent'")).toBeTruthy();
    expect(result.includes("export type myVector = 'activity'")).toBeTruthy();
    expect(result.includes("export type text_style = 'body'")).toBeTruthy();

    expect(result).toMatchSnapshot();
  });

  it('Should generate all the types with correct type name and formatting', async () => {
    const result = await toTypescriptDefinition(
      tokens as InputDataType,
      {
        formatName: 'snakeCase',
        typesNamesMapping: {
          color: 'another',
          opacity: 'name',
          shadow: 'why',
          measurement: 'not',
        },
      },
      libs,
    );

    expect(result.includes("export type another = 'colors_accent'")).toBeTruthy();
    expect(result.includes("export type name = 'subtle'")).toBeTruthy();
    expect(result.includes("export type why = 'elevation_1'")).toBeTruthy();
    expect(result.includes("export type not = 'base_space_01'")).toBeTruthy();

    expect(result).toMatchSnapshot();
  });

  it('Should generate all the types with name in pascalCase', async () => {
    const result = await toTypescriptDefinition(
      tokens as InputDataType,
      {
        formatName: 'pascalCase',
      },
      libs,
    );

    expect(result).toMatchSnapshot();
  });
});
