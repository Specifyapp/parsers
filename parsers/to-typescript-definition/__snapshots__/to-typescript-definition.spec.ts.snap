// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens to Typescript types definitions Should generate all the types with correct type name 1`] = `
"export type bitmap = 'acme-logo' | 'photo-example' | 'photo-example' | 'photo-example'

export type myVector = 'activity' | 'airplay' | 'alert-circle' | 'alert-octagon' | 'alert-triangle' | 'alert-triangle' | 'align-center' | 'align-center' | 'user-mask' | 'user-mask'

export type depth = 'background' | 'foreground' | 'middle'

export type duration = 'base' | 'long' | 'short' | 'very-long'

export type measurement = 'base-space-01' | 'base-space-02' | 'base-space-03' | 'base-space-04' | 'base-space-05' | 'base-space-06' | 'base-space-07' | 'base-space-08' | 'base-space-09' | 'base-space-10'

export type text_style = 'body' | 'body-with-opacity' | 'code' | 'list' | 'title'

export type border = 'border-accent' | 'border-accent-with-opacity' | 'border-accent-without-radii' | 'border-dashed'

export type helloColor = 'colors-accent' | 'colors-almost-black' | 'colors-black' | 'colors-green' | 'colors-grey' | 'colors-orange' | 'colors-pure-black' | 'colors-red' | 'colors-white'

export type shadow = 'elevation-1' | 'elevation-2' | 'elevation-3'

export type font = 'fira-code-medium' | 'inter-medium' | 'inter-semi-bold' | 'roboto-regular'

export type gradient = 'gradients-colored' | 'gradients-dark' | 'gradients-neutral' | 'gradients-safari'

export type opacity = 'subtle' | 'transparent' | 'visible'

"
`;

exports[`Tokens to Typescript types definitions Should generate all the types with correct type name and formatting 1`] = `
"export type bitmap = 'acme_logo' | 'photo_example' | 'photo_example' | 'photo_example'

export type vector = 'activity' | 'airplay' | 'alert_circle' | 'alert_octagon' | 'alert_triangle' | 'alert_triangle' | 'align_center' | 'align_center' | 'user_mask' | 'user_mask'

export type depth = 'background' | 'foreground' | 'middle'

export type duration = 'base' | 'long' | 'short' | 'very_long'

export type not = 'base_space_01' | 'base_space_02' | 'base_space_03' | 'base_space_04' | 'base_space_05' | 'base_space_06' | 'base_space_07' | 'base_space_08' | 'base_space_09' | 'base_space_10'

export type textStyle = 'body' | 'body_with_opacity' | 'code' | 'list' | 'title'

export type border = 'border_accent' | 'border_accent_with_opacity' | 'border_accent_without_radii' | 'border_dashed'

export type another = 'colors_accent' | 'colors_almost_black' | 'colors_black' | 'colors_green' | 'colors_grey' | 'colors_orange' | 'colors_pure_black' | 'colors_red' | 'colors_white'

export type why = 'elevation_1' | 'elevation_2' | 'elevation_3'

export type font = 'fira_code_medium' | 'inter_medium' | 'inter_semi_bold' | 'roboto_regular'

export type gradient = 'gradients_colored' | 'gradients_dark' | 'gradients_neutral' | 'gradients_safari'

export type name = 'subtle' | 'transparent' | 'visible'

"
`;

exports[`Tokens to Typescript types definitions Should generate all the types with default formatting 1`] = `
"export type bitmap = 'acme-logo' | 'photo-example' | 'photo-example' | 'photo-example'

export type vector = 'activity' | 'airplay' | 'alert-circle' | 'alert-octagon' | 'alert-triangle' | 'alert-triangle' | 'align-center' | 'align-center' | 'user-mask' | 'user-mask'

export type depth = 'background' | 'foreground' | 'middle'

export type duration = 'base' | 'long' | 'short' | 'very-long'

export type measurement = 'base-space-01' | 'base-space-02' | 'base-space-03' | 'base-space-04' | 'base-space-05' | 'base-space-06' | 'base-space-07' | 'base-space-08' | 'base-space-09' | 'base-space-10'

export type textStyle = 'body' | 'body-with-opacity' | 'code' | 'list' | 'title'

export type border = 'border-accent' | 'border-accent-with-opacity' | 'border-accent-without-radii' | 'border-dashed'

export type color = 'colors-accent' | 'colors-almost-black' | 'colors-black' | 'colors-green' | 'colors-grey' | 'colors-orange' | 'colors-pure-black' | 'colors-red' | 'colors-white'

export type shadow = 'elevation-1' | 'elevation-2' | 'elevation-3'

export type font = 'fira-code-medium' | 'inter-medium' | 'inter-semi-bold' | 'roboto-regular'

export type gradient = 'gradients-colored' | 'gradients-dark' | 'gradients-neutral' | 'gradients-safari'

export type opacity = 'subtle' | 'transparent' | 'visible'

"
`;

exports[`Tokens to Typescript types definitions Should generate all the types with name in pascalCase 1`] = `
"export type bitmap = 'AcmeLogo' | 'PhotoExample' | 'PhotoExample' | 'PhotoExample'

export type vector = 'Activity' | 'Airplay' | 'AlertCircle' | 'AlertOctagon' | 'AlertTriangle' | 'AlertTriangle' | 'AlignCenter' | 'AlignCenter' | 'UserMask' | 'UserMask'

export type depth = 'Background' | 'Foreground' | 'Middle'

export type duration = 'Base' | 'Long' | 'Short' | 'VeryLong'

export type measurement = 'BaseSpace01' | 'BaseSpace02' | 'BaseSpace03' | 'BaseSpace04' | 'BaseSpace05' | 'BaseSpace06' | 'BaseSpace07' | 'BaseSpace08' | 'BaseSpace09' | 'BaseSpace10'

export type textStyle = 'Body' | 'BodyWithOpacity' | 'Code' | 'List' | 'Title'

export type border = 'BorderAccent' | 'BorderAccentWithOpacity' | 'BorderAccentWithoutRadii' | 'BorderDashed'

export type color = 'ColorsAccent' | 'ColorsAlmostBlack' | 'ColorsBlack' | 'ColorsGreen' | 'ColorsGrey' | 'ColorsOrange' | 'ColorsPureBlack' | 'ColorsRed' | 'ColorsWhite'

export type shadow = 'Elevation1' | 'Elevation2' | 'Elevation3'

export type font = 'FiraCodeMedium' | 'InterMedium' | 'InterSemiBold' | 'RobotoRegular'

export type gradient = 'GradientsColored' | 'GradientsDark' | 'GradientsNeutral' | 'GradientsSafari'

export type opacity = 'Subtle' | 'Transparent' | 'Visible'

"
`;
